internal/defragger/defrag.go

```diff
@@ -0,0 +1,45 @@
+package defragger
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/improbable-eng/etcd-cluster-operator/internal/etcd"
+)
+
+type UsedSpaceChecker interface {
+	UsedSpacePercentage(context.Context, etcd.Member) (int, error)
+}
+
+type Defragger interface {
+	Defragment(context.Context, etcd.Member) error
+}
+
+// DefragIfNecessary will defrag each etcd member if necessary, it will return the last error encountered
+func DefragIfNecessary(ctx context.Context, memberList []etcd.Member, s UsedSpaceChecker, d Defragger, log logr.Logger) error {
+	var returnErr error
+	for _, member := range memberList {
+		usedSpace, err := s.UsedSpacePercentage(ctx, member)
+		if err != nil {
+			log.Error(err, "unable to check if defrag is necessary")
+			returnErr = err
+			continue
```

Comment:
Are we looking for the last error only? If yes, why not reverse the member list and look for the first?

Reply:
We try to do the defragging in the safest way, which means if we encounter an error defragging one member we still continue and try to defrag the other members. 

An issue occurs when we encounter an error when defragging multiple members, what should we do?
One solution is to return all errors we encounter. I chose to only return the most recent error, as it's the simplest solution.

Reply:
In my view hiding errors is a bad sign. I understand your reasons, the errors will be in the log, just like this one (because cron job doesn't have user interface). So it is up to you, it is just a feeling in me.